src/main/java/org/folio/rest/support/ShelvingOrderUpdate.java

```diff
@@ -0,0 +1,366 @@
+
+  public static int DEF_FETCH_SIZE = 5000;
+
+  private static ShelvingOrderUpdate instance;
+  private Versioned versionChecker;
+  private int rowsBunchSize;
+
+  public static ShelvingOrderUpdate getInstance() {
+    return getInstance(DEF_FETCH_SIZE);
+  }
+
+  public static ShelvingOrderUpdate getInstance(int rowsBunchSize) {
+    if (Objects.isNull(instance)) {
+      synchronized (ShelvingOrderUpdate.class) {
+        if (Objects.isNull(instance)) {
+          instance = new ShelvingOrderUpdate(rowsBunchSize);
+          log.info("An shelving oder updater instantiated...");
+        }
+      }
+    }
+    return instance;
+  }
+
+  private ShelvingOrderUpdate(int rowsBunchSize) {
+    this.versionChecker = new Versioned() {};
+    this.versionChecker.setFromModuleVersion(VERSION_WITH_SHELVING_ORDER);
+    this.rowsBunchSize = rowsBunchSize;
+  }
+
+  /**
+   * Entry routine of items updates. Invokes the process recursively until no items to update
+   *
+   * @param attributes
+   * @param headers
+   * @param vertxContext
+   * @return
+   */
+  public Future<Integer> startUpdatingOfItems(TenantAttributes attributes, Map<String, String> headers,
+      Context vertxContext) {
+
+    final Vertx vertx = vertxContext.owner();
+
+    log.info("Update items started, moduleTo is: \"{}\",  moduleFrom is: \"{}\"",
+        attributes.getModuleTo(), attributes.getModuleFrom());
+
+    if (! ShelvingOrderUpdate.getInstance().isAllowedToUpdate(attributes)) {
+      log.info("Module isn't eligible for upgrade, just exit.");
+      // Not allowed to perform items updates, just returns
+      return Future.succeededFuture(0);
+    }
+
+    log.info("Module is eligible for upgrade.");
+    // Updates items with shelving order property.
+    // This should be triggered once for particular version only.
+    PostgresClient postgresClient = PgUtil.postgresClient(vertxContext, headers);
+
+    final long startTime = System.currentTimeMillis();
+
+    return fetchAndUpdateAll(vertx, postgresClient)
+        .onComplete(x -> log.info("Items updates with shelving order property completed in {} milliseconds",
+            (System.currentTimeMillis() - startTime)));
+  }
+
+  private Future<Integer> fetchAndUpdateAll(Vertx vertx, PostgresClient postgresClient) {
+    return fetchAndUpdateBunch(vertx, postgresClient)
+    .compose(headRows -> {
+      if (headRows <= 0) {
+        return Future.succeededFuture(0);
+      }
+      return fetchAndUpdateAll(vertx, postgresClient).map(tailRows -> headRows + tailRows);
+    });
+  }
+
+  /**
+   * Performs fetching items by fixed bunches of rows with further property calculations and
+   * updating in database
+   *
+   * @param vertx
+   * @param postgresClient
+   * @return
+   */
+  private Future<Integer> fetchAndUpdateBunch(Vertx vertx, PostgresClient postgresClient) {
+    log.info("The routine of \"fetchAndUpdatesItems\" has started");
+    Promise<Integer> fetchAndUpdatePromise = Promise.promise();
```

Comment:
Consider avoiding this Promise.
`return postgresClient.getConnection().compose(connection -> {`
or
`return postgresClient.withTransaction(tx -> {`
You may need to add `.map(x -> someReturnValue)` at the end.

Reply:
Or `.mapEmpty()`.

Reply:
The point here is to place all the steps in single transaction related execution instead of spawning a new transaction with further needing to control it. Correct me if I'm wrong, please.

Reply:
Thanks, the new code is good!